ndb/src/google/cloud/ndb/query.py

```diff
@@ -129,43 +140,729 @@ def resolve(self, bindings, used):
+    datastore, for example a query for a structured value.
+
+    Args:
+        predicate (Callable[[Any], bool]): A filter predicate that
+            takes a datastore entity (typically as a protobuf) and
+            returns :data:`True` or :data:`False` if the entity matches
+            the given filter.
+    """
+
+    __slots__ = ("predicate",)
+
+    def __new__(cls, predicate):
+        self = super(PostFilterNode, cls).__new__(cls)
+        self.predicate = predicate
+        return self
+
+    def __getnewargs__(self):
+        """Private API used to specify ``__new__`` arguments when unpickling.
+
+        .. note::
+
+            This method only applies if the ``pickle`` protocol is 2 or
+            greater.
+
+        Returns:
+            Tuple[Callable[[Any], bool],]: A tuple containing a single value,
+            the ``predicate`` attached to this node.
+        """
+        return (self.predicate,)
+
+    def __repr__(self):
+        return "{}({})".format(self.__class__.__name__, self.predicate)
+
+    def __eq__(self, other):
+        if not isinstance(other, PostFilterNode):
+            return NotImplemented
+        return self is other or self.predicate == other.predicate
+
+    def _to_filter(self, post=False):
+        """Helper to convert to low-level filter, or :data:`None`.
+
+        Args:
+            post (bool): Indicates if this is a post-filter node.
+
+        Returns:
+            Tuple[Callable[[Any], bool], None]: If this is a post-filter, this
+            returns the stored ``predicate``, otherwise it returns
+            :data:`None`.
+        """
+        if post:
+            return self.predicate
+        else:
+            return None
+
+
+class _BooleanClauses:
+    """This type will be used for symbolically performing boolean operations.
+
+    Internally, the state will track a symbolic expression like::
+
+        A or (B and C) or (A and D)
+
+    as a list of the ``OR`` components::
+
+        [A, B and C, A and D]
+
+    When ``combine_or=False``, it will track ``AND`` statements as a list,
+    making the final simplified form of our example::
+
+        [[A], [B, C], [A, D]]
+
+    Via :meth:`add_node`, we will ensure that new nodes will be correctly
+    combined (via ``AND`` or ``OR``) with the current expression.
+
+    Args:
+        name (str): The name of the class that is tracking a
+            boolean expression.
+        combine_or (bool): Indicates if new nodes will be combined
+            with the current boolean expression via ``AND`` or ``OR``.
+    """
+
+    __slots__ = ("name", "combine_or", "or_parts")
+
+    def __init__(self, name, combine_or):
+        self.name = name
+        self.combine_or = combine_or
+        if combine_or:
+            # For ``OR()`` the parts are just nodes.
+            self.or_parts = []
+        else:
+            # For ``AND()`` the parts are "segments", i.e. node lists.
+            self.or_parts = [[]]
```

Comment:
The type flag smells odd, as thought there should really be two separate classes.

Reply:
That's probably a good idea. The `OR` case is significantly simpler, to where it doesn't really need a class. WDYT? Should I have a `_AndClauses` and `_OrClauses` just for consistency?